src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -18,43 +22,82 @@ public DefaultObjectPool(IPooledObjectPolicy<T> policy)
 
         public DefaultObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
         {
-            if (policy == null)
+            _policy = policy ?? throw new ArgumentNullException(nameof(policy));
+            _isDefaultPolicy = IsDefaultPolicy();
+
+            // -1 due to _firstItem
+            _items = new ObjectWrapper[maximumRetained - 1];
+
+            bool IsDefaultPolicy()
             {
-                throw new ArgumentNullException(nameof(policy));
-            }
+                var type = policy.GetType();
 
-            _policy = policy;
-            _items = new T[maximumRetained];
+                return type.IsGenericType
+                    ? type.GetGenericTypeDefinition() == typeof(DefaultPooledObjectPolicy<>)
+                    : false;
+            }
         }
 
         public override T Get()
         {
-            for (var i = 0; i < _items.Length; i++)
+            T item = _firstItem;
+
+            if (item == null || Interlocked.CompareExchange(ref _firstItem, null, item) != item)
             {
-                var item = _items[i];
-                if (item != null && Interlocked.CompareExchange(ref _items[i], null, item) == item)
+                item = GetViaScan();
+            }
+
+            return item;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T GetViaScan()
+        {
+            ObjectWrapper[] items = _items;
+            T item = null;
+
+            for (int i = 0; i < items.Length; i++)
+            {
+                item = items[i];
+
+                if (item != null && Interlocked.CompareExchange(ref items[i].Element, null, item) == item)
                 {
-                    return item;
+                    break;
                 }
             }
 
-            return _policy.Create();
+            return item ?? _policy.Create();
```

Comment:
@gfoidl - to what degree is interface dispatch a significant contributor? We could consider adding an abstract class for the policy and then pivoting the implementation to primarily use the class.

In the case of the default policy this means it would be a virtual method call vs an iface call. 

If you don't have data or the interest to pursue that then that's fine 😆 your change is a big improvement on its own.

Reply:
I'll investigate this, and post the results here.

Reply:
The call to `_policy.Create` is only done, when no "usable" object in the pool was found. So it is on the absolut cold path. Hence the interface call should be no problem.

Anyway I'll do some benchmarks on this.

Reply:
I tried something with the abstract base class -- https://github.com/gfoidl/Common/commit/e5dfc679f908db64b1a8302d0d9d1c8410a8469f

As stated in the [previous comment](https://github.com/aspnet/Common/pull/314#discussion_r169399066) the call to `Create` is on the cold path, so the cost of interfavce vs. abstract class is negligible and the measurable difference lies within the noise.

I haven't looked more in detail at https://github.com/aspnet/Common/blob/41406a403a99e518d8f5a32007fe84c017ebeed1/src/Microsoft.Extensions.ObjectPool/StringBuilderPooledObjectPolicy.cs#L19 but there might be a noticeable difference.

Tomorrow I'll try something to devirtualize the calls to Create and Return, so especially for the StirngBuilder the perf-win should be significant.

Maybe it is better to handle this in a new PR?!

Reply:
> Maybe it is better to handle this in a new PR?! 

Sure

Reply:
Regarding the question about the interface call:

``` ini

BenchmarkDotNet=v0.10.12, OS=Windows 10 Redstone 3 [1709, Fall Creators Update] (10.0.16299.248)
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=2742187 Hz, Resolution=364.6724 ns, Timer=TSC
.NET Core SDK=2.1.4
  [Host]     : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.5 (Framework 4.6.26020.03), 64bit RyuJIT


```
|         Method |     Mean |     Error |    StdDev | Scaled | ScaledSD |
|--------------- |---------:|----------:|----------:|-------:|---------:|
|         Direct | 1.482 ns | 0.0239 ns | 0.0223 ns |   1.00 |     0.00 |
|      Interface | 2.205 ns | 0.0150 ns | 0.0140 ns |   1.49 |     0.02 |
|       Abstract | 1.533 ns | 0.0160 ns | 0.0150 ns |   1.03 |     0.02 |
| AbstractSealed | 1.543 ns | 0.0079 ns | 0.0074 ns |   1.04 |     0.02 |
|       Delegate | 1.853 ns | 0.0088 ns | 0.0082 ns |   1.25 |     0.02 |
|     Expression | 3.918 ns | 0.0275 ns | 0.0243 ns |   2.64 |     0.04 |
|    IlGenSimple | 1.866 ns | 0.0131 ns | 0.0123 ns |   1.26 |     0.02 |
|          IlGen | 1.831 ns | 0.0233 ns | 0.0218 ns |   1.24 |     0.02 |

[Code](https://github.com/gfoidl/Benchmarks/blob/aeb07ce692215abeff241eba5e04f51404c2486a/misc/CallTimes/CallTimes/Program.cs)